src/main/java/emissary/core/MobileAgent.java

```diff
@@ -526,6 +526,11 @@ protected DirectoryEntry getNextKey(@Nullable final IServiceProviderPlace place,
                 String formID = form + KeyManipulator.DATAIDSEPARATOR + stageName;
                 curEntry = nextKeyFromDirectory(formID, place, lastEntry, payloadArg);
 
+                // if last place/directory entry is denied, reject
+                if (curEntry != null && place.isDenied(curEntry.getDataType())) {
```

Comment:
This might be closer to what we need, although I have a feeling we really need the check further upstream so a denied place will not be returned by the `nextKeyFromDirectory` call:

```java
                // if next proposed place/directory entry is denied, reject
                if (curEntry != null && curEntry.getLocalPlace().isDenied(form)) {
                    logger.warn("FORM {} is denied for place {}; continuing", form, curEntry.getLocalPlace().getPlaceName());
                    continue;
                }
```